File: StockExchangeServer/Services/StockExchangeService.cs
using Grpc.Core;
using StockExchangeServer;
using StockExchangeServer.Models;

namespace StockExchangeServer.Services
{
    public class StockExchangeService : StockExchange.StockExchangeBase
    {
        private readonly ILogger<StockExchangeService> _logger;
        private readonly StockMarket _market;

        public StockExchangeService(ILogger<StockExchangeService> logger, StockMarket market)
        {
            _logger = logger;
            _market = market;
        }

        public override async Task SubscribeQuotes(QuoteSubscription request, IServerStreamWriter<StockQuote> responseStream, ServerCallContext context)
        {
            _logger.LogInformation($"Клиент подписался на данные по символу: {string.Join(", ", request.Symbols)}|");


            var symbols = request.Symbols.Count == 0 ? _market.GetAvailableSymbols() : request.Symbols.ToList<string>();

            try {
                while (!context.CancellationToken.IsCancellationRequested)
                {
                    foreach(var symbol in symbols)
                    {
                        var quote = _market.GetQuote(symbol); // Получаем котировку
                        if (quote != null)
                        {
                            await responseStream.WriteAsync(quote);
                        }
                    }
                    await Task.Delay(request.UpdateIntervalMs > 0 ? request.UpdateIntervalMs : 1000, context.CancellationToken);

                }
            }
            catch(OperationCanceledException) {   // Обработка отмены операции
                _logger.LogInformation("Подписка на котироку была отменена");
            }

        }

        public override Task<OrderResponse> PlaceOrder(OrderRequest request, ServerCallContext context)
        {
            _logger.LogInformation($"Размещение ордера: {request.OrderId} для клиента {request.ClientId}");

            var response = _market.PlaceOrder(request); // Разместить ордер через модель биржи

            return Task.FromResult(response);   // Возвращаем рузультат как завершунную задачу
        }


        public override Task<CancelOrderResponse> CancelOrder(CancelOrderRequest request, ServerCallContext context)
        {
            _logger.LogInformation($"Отмена ордера: {request.OrderId} для клиента {request.ClientId}");
            var response = _market.CancelOrder(request.OrderId, request.ClientId); // Отмена ордера через модель биржи

            return Task.FromResult(response);
        }



        public override Task<OrderHistoryResponse> GetOrderHistory(OrderHistoryRequest request, ServerCallContext context) {
            var orders = _market.GetOrderHistory(request.ClientId, request.MaxOrders); // Получаем историю ордеров
            return Task.FromResult(new OrderHistoryResponse { Orders = { orders } });
        }

        public override Task<PortfolioResponse> GetPortfolio(PortfolioRequest request, ServerCallContext context)
        {
            var portfolio = _market.GetPortfolio(request.ClientId); // Получаем портфель клиента
            return Task.FromResult(portfolio);
        }











        /*****************************/

    }

}

File: stock.exchange.react/src/services/stockExchangeService.ts
import * as grpcWeb from 'grpc-web';
import { StockExchangeClient } from '../types/StockServiceClientPb';
import {
  QuoteSubscription,
  StockQuote,
  OrderRequest,
  OrderResponse,
  CancelOrderRequest,
  CancelOrderResponse,
  OrderHistoryRequest as OrderHistoryRequestMessage,
  OrderHistoryResponse as OrderHistoryResponseMessage,
  PortfolioRequest as PortfolioRequestMessage,
  PortfolioResponse as PortfolioResponseMessage,
} from '../types/stock_pb';
import { Guid } from 'guid-ts';

class StockExchangeService {
  private readonly client: StockExchangeClient;
  private readonly hostname: string = 'http://localhost:50051'; // Assuming grpc-web proxy is running here

  constructor() {
    this.client = new StockExchangeClient(this.hostname);
  }

  subscribeQuotes(
    symbols: string[],
    onMessage: (quote: StockQuote) => void,
    onError: (error: grpcWeb.RpcError) => void,

    updateIntervalMs: number = 2000
  ): grpcWeb.ClientReadableStream<StockQuote> {
    const request = new QuoteSubscription();
    request.setSymbolsList(symbols);
    request.setUpdateIntervalMs(updateIntervalMs);

    const stream = this.client.subscribeQuotes(request, {});

    stream.on('data', onMessage);
    stream.on('error', onError);


    return stream;
  }

  async placeOrder(orderRequest: OrderRequest): Promise<OrderResponse> {
    const request = new OrderRequest();
    request.setOrderId(`order_${Guid.newGuid()}`);
    request.setSymbol(orderRequest.getSymbol());
    request.setOrderType(orderRequest.getOrderType());
    request.setDirection(orderRequest.getDirection());
    request.setQuantity(orderRequest.getQuantity());
    request.setClientId(orderRequest.getClientId());
    request.setPrice(orderRequest.getPrice());
    request.setTimeInForce(orderRequest.getTimeInForce());

    return this.client.placeOrder(request, null);
  }

  async cancelOrder(
    orderId: string,
    clientId: string
  ): Promise<CancelOrderResponse> {
    const request = new CancelOrderRequest();
    request.setOrderId(orderId);
    request.setClientId(clientId);

    return this.client.cancelOrder(request, null);
  }

  async getOrderHistory(
    clientId: string,
    maxOrders: number = 10
  ): Promise<OrderHistoryResponseMessage> {
    const request = new OrderHistoryRequestMessage();
    request.setClientId(clientId);
    request.setMaxOrders(maxOrders);

    return this.client.getOrderHistory(request, null);
  }

  async getPortfolio(
    clientId: string
  ): Promise<PortfolioResponseMessage> {
    const request = new PortfolioRequestMessage();
    request.setClientId(clientId);

    return this.client.getPortfolio(request, null);
  }
}

export const stockExchangeService = new StockExchangeService();